jwst/background/background_sub.py

```diff
@@ -91,3 +109,191 @@ def average_background(bkg_list):
 
     return avg_bkg
+
+
+def subtract_wfss_bkg(input_model, bkg_ref):
+    """Scale and subtract a background reference image from WFSS/GRISM data.
+
+    Parameters
+    ----------
+    input_model: JWST data model
+        input target exposure data model
+
+    bkg_ref: reference file model
+        master background file for WFSS/GRISM
+
+    Returns
+    -------
+    result: JWST data model
+        background-subtracted target data model
+    """
+
+    if hasattr(input_model.meta, "source_catalog"):
+        source_catalog = input_model.meta.source_catalog.filename
+        got_catalog = True
+    else:
+        log.warning("No source_catalog found in input.meta.")
+        got_catalog = False
+
+    # If there are NaNs, we have to replace them with something harmless.
+    bkg_ref = no_NaN(bkg_ref)
+
+    # Create a mask from the source catalog, True where there are no sources,
+    # i.e. in regions we can use as background.
+    if got_catalog:
+        bkg_mask = mask_from_source_cat(input_model.data.shape, source_catalog)
+    else:
+        bkg_mask = np.ones(input_model.data.shape, dtype=np.bool)
+
+    # Compute the mean values of science image and background reference
+    # image, including only regions where there are no identified sources.
+    # Exclude pixel values in the lower and upper 25% of the histogram.
+    lowlim = 25.
+    highlim = 75.
+    sci_mean = robust_mean(input_model.data[bkg_mask],
+                           lowlim=lowlim, highlim=highlim)
+    bkg_mean = robust_mean(bkg_ref.data[bkg_mask],
+                           lowlim=lowlim, highlim=highlim)
+
+    log.debug("mean of [{}, {}] percentile science data = {}"
+              .format(lowlim, highlim, sci_mean))
+    log.debug("mean of [{}, {}] percentile background image = {}"
+              .format(lowlim, highlim, bkg_mean))
+
+    result = input_model.copy()
+    if bkg_mean != 0.:
+        subtract_this = (sci_mean / bkg_mean) * bkg_ref.data
+        result.data = input_model.data - subtract_this
+        log.debug("Average of values subtracted = {}"
+                  .format(subtract_this.mean(dtype=np.float)))
+    else:
+        log.warning("Background file has zero mean; "
+                    "nothing will be subtracted.")
+    result.dq = np.bitwise_or(input_model.dq, bkg_ref.dq)
+    return result
+
+
+def no_NaN(model, fill_value=0.):
+    """Replace NaNs with a harmless value.
+
+    Parameters
+    ----------
+    model: JWST data model
+        Reference file model.
+
+    fill_value: float
+        NaNs will be replaced with this value.
+
+    Returns
+    -------
+    result: JWST data model
+        Reference file model without NaNs in data array.
+    """
+
+    mask = np.isnan(model.data)
+    if mask.sum(dtype=np.intp) == 0:
+        return model
+    else:
+        temp = model.copy()
+        temp.data[mask] = fill_value
+        return temp
+
+
+def mask_from_source_cat(shape, source_catalog):
```

Comment:
The mask for grism mode cannot be made directly from the source catalog. The source catalog *only* contains information about the objects in the direct image. We must use the list of GrismObjects which contain the translated bounding boxes which should be used to avoid areas in the sky background reference image.